### PR TITLE
Allow forking snapshot more than 1 time

### PR DESCRIPTION
**Pull request by yaroslava-serdiuk**
#### Which component this PR applies to?

cluster-autoscaler

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Forking twice is needed for parallel scale down implementation.

### CODE REVIEW

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go
Comment (x13n) [inline cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go:58]: Why impose an arbitrary limit?

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot_test.go
Comment (x13n) [inline cluster-autoscaler/simulator/clustersnapshot/clustersnapshot_test.go:217]: I think there should be more test cases like this: fork & fork & commit & revert, fork & fork & revert & commit, with snapshot changes between these steps.

File: cluster-autoscaler/core/scaledown/planner/planner.go
Comment (x13n) [inline cluster-autoscaler/core/scaledown/planner/planner.go:88]: `RemovalSimulator` operates on ClusterSnapshot passed through context. Since we have access to `p.context` here, I'd just fork it here instead of extending `removalSimulator` interface. Also note that when we fork, we also need to `defer Revert()` afterwards.

File: cluster-autoscaler/core/scaledown/planner/planner.go
Comment (x13n) [inline cluster-autoscaler/core/scaledown/planner/planner.go:89]: Forking the snapshot shouldn't return an error anymore. If forking can happen multiple times, `Fork()` shouldn't return a value at all. This will simplify error handling in a few places.

**Review by x13n — CHANGES_REQUESTED** [review]

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go:58]: The main reason why there was a limitation to fork once previously is to ensure we don't accidentally fork twice.
The main reason why I added a fork limit is to control recursion depth of forking. Currently we limit fork to two time, so I used this value as a new default.

File: cluster-autoscaler/core/scaledown/planner/planner.go
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/core/scaledown/planner/planner.go:89]: >Forking the snapshot shouldn't return an error anymore
Why? I added a fork limit, so the Fork() can return error.

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go
Comment (x13n) [reply cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go:58]: I disagree handling this as a runtime error is the right approach. If we are concerned there are parts of the logic which could fork without committing/reverting, there should be unit tests ensuring this doesn't happen.

Comment (x13n) [reply cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go:58]: Alternatively, we could modify the interface in a way that ensures forking without commiting/reverting is not possible at all. I actually did just that in one place already:
https://github.com/kubernetes/autoscaler/blob/a703c3f6cff1d54dc92378c5baad2c2cccdbbc8f/cluster-autoscaler/simulator/cluster.go#L198-L217

File: cluster-autoscaler/core/scaledown/planner/planner.go
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/core/scaledown/planner/planner.go]: Done

File: cluster-autoscaler/core/scaledown/planner/planner.go
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/core/scaledown/planner/planner.go]: Ack, thanks!

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot_test.go
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/simulator/clustersnapshot/clustersnapshot_test.go:209]: Done

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go]: Done

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot_test.go
Comment (x13n) [inline cluster-autoscaler/simulator/clustersnapshot/clustersnapshot_test.go:100]: Looks like this just duplicates existing test case below, why add a second one?

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot_test.go
Comment (x13n) [inline cluster-autoscaler/simulator/clustersnapshot/clustersnapshot_test.go:755]: Let's not depend on filtering out specific test cases like this. Why test just one?

**Review by x13n — CHANGES_REQUESTED** [review]

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot_test.go
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/simulator/clustersnapshot/clustersnapshot_test.go:755]: Oops, sorry it's a leftover from debugging, thanks for catching it!

File: cluster-autoscaler/simulator/clustersnapshot/basic.go
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/simulator/clustersnapshot/basic.go:26]: Removing it since it sounds confusing. My intend is to describe that now basic snapshot is a list of forks.

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go:58]: I see your point. However now we allow fork infinite time, which actually I would like to avoid somehow.

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot_benchmark_test.go
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/simulator/clustersnapshot/clustersnapshot_benchmark_test.go:205]: Seems I wanted to modify the test and fork twice, but I forgot to fork second time :)

File: cluster-autoscaler/simulator/clustersnapshot/basic.go
Comment (x13n) [reply cluster-autoscaler/simulator/clustersnapshot/basic.go]: Got it. I think this is an implementation detail, which doesn't have to be a part of object doc comment.

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go
Comment (x13n) [reply cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go:58]: Sure, we can document that `Fork`, `Commit` and `Revert` are not meant to be invoked directly. It is always up to the person implementing changes to write correct code - giving them the tool to do so easily (the helper function) seems sufficient to me.

**Comment by x13n** [comment]
Thanks for the change, LGTM now. I'd still drop that TODO, but don't want to block on this. Putting on hold to let you make the change or just drop the hold if you still disagree.

/lgtm
/hold

File: cluster-autoscaler/simulator/clustersnapshot/basic.go
Comment (MaciekPytel) [inline cluster-autoscaler/simulator/clustersnapshot/basic.go:278]: nit: I think direct pop from the list would be much more readable, I found this very confusing when skimming through the code (I'm not saying this is incorrect, just surprising to me).

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go
Comment (MaciekPytel) [inline cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go:61]: I don't really like using error returned by passing function as a signal to commit/revert. 

I think the decision to commit or revert depends more on what the forked snapshot is used for (scale-up/scale-down simulation vs updating the expected state of the cluster) and not necessarily on the success / failure of the logic. Or, in other words, just reverting a forked snapshot and proceeding is generally not the error handling strategy for a failed simulation. By using `func() error` we have no convenient way of propagating the error outside of WithForkedSnapshot.

I'd much rather this took `func() (bool, error)` where the boolean return value determines if the forked snapshot should be commited or reverted. This would separate the decision from any other error handling and force the user to make explicit decision on whether the snapshot should be committed or reverted for each possible error condition.

**Comment by MaciekPytel** [comment]
/approve
/hold
Holding to give you a chance to look at my comments before this is merged.

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go]: Agreed, thanks for pointing this out! I modified the function and changed the error returned from WithForkedSnapshot regarding cluster snapshot operation. I think we should return error that shows if expected operation was completed (Commit or Revert). 
However all of the implementations of Revert() return nil. And I suppose since Revert() is direct opposite to Fork(), so they both should return the same value. So I actually think that we can drop returned value for Revert() as we did it for Fork(). WDYT?

Comment (MaciekPytel) [reply cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go]: I guess if the revert cannot fail in any implementation, I think we can drop the error from signature. I don't really expect implementations of this interface outside the ones we have, so changing interface doesn't seem like an issue.
The only condition I can think of when revert could fail is if the snapshot was not forked, but IIRC we just treat this as no-op and don't return an error?

Comment (yaroslava-serdiuk) [reply cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go]: Yes, if we haven't fork before calling revert we just do nothing and return nil

Comment (yaroslava-serdiuk) [reply cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go]: Dropped error for Revert()

File: cluster-autoscaler/simulator/clustersnapshot/basic.go
Comment (MaciekPytel) [inline cluster-autoscaler/simulator/clustersnapshot/basic.go:276]: Why not just `snapshot.data = snapshot.data[:len(snapshot.data)-1]` like you do in revert? Is there some subtle difference in behavior between the two?

Comment (yaroslava-serdiuk) [reply cluster-autoscaler/simulator/clustersnapshot/basic.go:276]: My understanding of commit() is following: if we have a chain of snapshots A -> B -> C -> D (D is the last fork), then after commit we would like to have A -> B -> D. So we want to have everything up to len-2 (excluding) + the last element. 
Reverting is definitely what we want to have as result for commit()

Comment (MaciekPytel) [reply cluster-autoscaler/simulator/clustersnapshot/basic.go:276]: Oh, right, we're deleting last-but-one element. Makes perfect sense, sorry, my comment makes no sense :(

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go]: Done

File: cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go
Comment (yaroslava-serdiuk) [reply cluster-autoscaler/simulator/clustersnapshot/clustersnapshot.go]: Right, change Fatalf to Errorf, since we can revert and nothing bad will happen.

**Pull request merged**
